hudi-client/hudi-spark-client/src/main/java/org/apache/hudi/index/bucket/HoodieSparkConsistentBucketIndex.java

```diff
@@ -0,0 +1,218 @@
+      ConsistentBucketIdentifier identifier = new ConsistentBucketIdentifier(metadata);
+      partitionToIdentifier.put(p, identifier);
+    });
+  }
+
+  /**
+   * Get bucket location for given key and partition
+   *
+   * @param key
+   * @param partitionPath
+   * @return
+   */
+  @Override
+  protected HoodieRecordLocation getBucket(HoodieKey key, String partitionPath) {
+    ConsistentHashingNode node = partitionToIdentifier.get(partitionPath).getBucket(key, indexKeyFields);
+    if (node.getFileIdPfx() != null && !node.getFileIdPfx().isEmpty()) {
+      /**
+       * Dynamic Bucket Index doesn't need the instant time of the latest file group.
+       * We add suffix 0 here to the file uuid, following the naming convention.
+       */
+      return new HoodieRecordLocation(null, FSUtils.createNewFileId(node.getFileIdPfx(), 0));
+    }
+
+    LOG.error("Consistent hashing node has no file group, partition: " + partitionPath + ", meta: "
+        + partitionToIdentifier.get(partitionPath).getMetadata().getFilename() + ", record_key: " + key.toString());
+    throw new HoodieIndexException("Failed to getBucket as hashing node has no file group");
+  }
+
+  /**
+   * Load hashing metadata of the given partition, if it is not existed, create a new one (also persist it into storage)
+   *
+   * @param table     hoodie table
+   * @param partition table partition
+   * @return Consistent hashing metadata
+   */
+  public HoodieConsistentHashingMetadata loadOrCreateMetadata(HoodieTable table, String partition) {
+    int retry = 3;
+    // TODO maybe use ConsistencyGuard to do the retry thing
+    // retry to allow concurrent creation of metadata (only one attempt can succeed)
+    while (retry-- > 0) {
+      HoodieConsistentHashingMetadata metadata = loadMetadata(table, partition);
+      if (metadata == null) {
+        metadata = new HoodieConsistentHashingMetadata(partition, numBuckets);
+        if (saveMetadata(table, metadata, false)) {
+          return metadata;
+        }
+      } else {
+        return metadata;
+      }
+    }
+    throw new HoodieIndexException("Failed to load or create metadata, partition: " + partition);
+  }
+
+  /**
+   * Load hashing metadata of the given partition, if it is not existed, return null
+   *
+   * @param table     hoodie table
+   * @param partition table partition
+   * @return Consistent hashing metadata or null if it does not exist
+   */
+  public static HoodieConsistentHashingMetadata loadMetadata(HoodieTable table, String partition) {
+    Path metadataPath = FSUtils.getPartitionPath(table.getMetaClient().getHashingMetadataPath(), partition);
+
+    try {
+      if (!table.getMetaClient().getFs().exists(metadataPath)) {
+        return null;
+      }
+      FileStatus[] metaFiles = table.getMetaClient().getFs().listStatus(metadataPath);
+      HoodieTimeline completedCommits = table.getMetaClient().getActiveTimeline().getCommitTimeline().filterCompletedInstants();
```

Comment:
What about compactions? Can the index be updated after compaction?

Reply:
In the current design, index (metadata) of a partition will be created in the first write and be updated in clustering (split&merge bucket). Compaction will not modify the metadata and is exclusive with clustering.